vaadin-text-field-flow-parent/vaadin-text-field-flow/src/main/java/com/vaadin/flow/component/textfield/AbstractNumberField.java

```diff
@@ -138,11 +138,19 @@ public void setValue(T value) {
 
         super.setValue(value);
 
+        // Clear the input element from possible bad input.
         if (Objects.equals(oldValue, getEmptyValue())
                 && Objects.equals(value, getEmptyValue())
                 && isInputValuePresent()) {
-            // Clear the input element from possible bad input.
-            getElement().executeJs("this.inputElement.value = ''");
+            // The check for value presence guarantees that a non-empty value
+            // won't get cleared when setValue(null) and setValue(...) are
+            // subsequently called within one round-trip.
+            // That may happen otherwise due to the Flow optimization:
+            // Flow only sends the final component value to the client
+            // when you update the value multiple times during a round-trip
+            // and the final value is sent in place of the first one.
+            getElement()
+                    .executeJs("if (!this.value) this._inputElementValue = ''");
             getElement().setProperty("_hasInputValue", false);
```

Comment:
This PR looks Ok to me, I only wonder if ` getElement().setProperty("_hasInputValue", false);` shouldn't get a similar treatment. However, I'm not sure if I understand what `_hasInputValue` should express: I thought simply that _input has some value in it_, but when I set a value to a field using flow's `integerField.setValue(1234)` it is still `false`. It changes to `_hasInputValue==true` only when I set the input manually by typing text in the field... is that correct?

Reply:
Yeah, it is a known inconsistency that `_hasInputValue` can indicate `false` when the value is set programmatically. This currently doesn't cause real issues because Flow relies on this flag when it needs to determine whether _the user_ has entered something in the field. The inconsistency is described in: https://github.com/vaadin/web-components/issues/5639.

The `_hasInputValue` property isn't prone to this regression because we are using `setProperty`. Only the final value is sent to the client.